cpp/ql/src/semmle/code/cpp/models/implementations/Printf.qll

```diff
@@ -153,58 +172,120 @@ class Snprintf extends FormattingFunction {
     not exists(getDefinition().getFile().getRelativePath())
   }
 
-  override int getSizeParameterIndex() {
-    result = 1
+  override int getSizeParameterIndex() { result = 1 }
+}
+
+/**
+ * The standard functions `vprintf` and `vwprintf`, and their
+ * assorted variants.
+ */
+class Vprintf extends FormattingFunction {
+  Vprintf() {
+    this instanceof TopLevelFunction and
+    (
+      hasGlobalName("vprintf") or
+      hasGlobalName("__builtin_vprintf") or
+      hasGlobalName("vfprintf") or
+      hasGlobalName("__builtin_vfprintf") or
+      hasGlobalName("vsprintf") or
+      hasGlobalName("__builtin_vsprintf") or
+      hasGlobalName("vsnprintf") or
+      hasGlobalName("__builtin_vsnprintf") or
+      hasGlobalName("vprintf_s") or
+      hasGlobalName("vfprintf_s") or
+      hasGlobalName("vsprintf_s") or
+      hasGlobalName("vsnprintf_s") or
+      hasGlobalName("_vsnprintf_s") or
+      hasGlobalName("_vsnprintf_s_l") or
+      hasGlobalName("vwprintf") or
+      hasGlobalName("vfwprintf") or
+      hasGlobalName("vswprintf") or
+      hasGlobalName("vwprintf_s") or
+      hasGlobalName("vfwprintf_s") or
+      hasGlobalName("vswprintf_s") or
+      hasGlobalName("_vsnwprintf_s") or
+      hasGlobalName("_vsnwprintf_s_l")
+    ) and
+    not exists(getDefinition().getFile().getRelativePath())
+  }
+
+  override int getFormatParameterIndex() {
+    if getName().matches("%\\_l")
+    then result = getFirstFormatArgumentIndex() - 2
+    else result = getFirstFormatArgumentIndex() - 1
   }
+
+  override int getFirstFormatArgumentIndex() { result = getNumberOfParameters() - 1 }
```

Comment:
Perhaps this predicate should be `none()`, with qldoc explaining that a result is not appropriate because the format arguments are not explicitly provided for `vprintf` in the same way as for a `printf`?
e.g.
```
printf("%i %i", 1, 2); // getFirstFormatArgumentIndex() = 1
vprintf("%i %i", args); // has no getFirstFormatArgumentIndex() ?
```
I suspect this change will prevent existing queries such as `WrongTypeFormatArguments.ql` from producing unexpected results.